### ISSUE TITLE
Add `conv2d_transpose` method to `theano.tensor.nnet.abstract_conv`

### ISSUE THREAD

**Issue by lamblin** [Theano/Theano#3989]
... and make it available in `theano.tensor.nnet`.
The semantics and name of the parameters was discussed IRL with @vdumoulin and @fvisin .

This is a follow-up to #3681 and #3872.

**Comment by meanmee** [comment]
I used the code below to install theano：
`pip install git+git://github.com/Theano/Theano.git`
But there is no thenao.tensor.nnet.conv2d_transpose() method.

**Comment by fvisin** [comment]
IIRC you should create a new `conv2d_transpose` op in `theano.tensor.nnet.abstract_conv`, which will be an alias of `theano.tensor.nnet.abstract_conv.conv2d_grad_wrt_inputs` with more user-friendly parameter names. Then, you should make this new op available in `theano.tensor.nnet`.

**Comment by lamblin** [comment]
@meanmee: this is an open issue, meaning it is not done yet. Someone has been assigned to it, though, so it should become available soon.

**Comment by fvisin** [comment]
Oh sorry, I thought you had been assigned to this issue.

**Comment by vdumoulin** [comment]
In order to design the interface of **Theano**'s `conv2d_transpose` op, I had a look at **TensorFlow**'s `conv2d_transpose` interface:

#### `tf.nn.conv2d_transpose(value, filter, output_shape, strides, padding='SAME', name=None)`

* `value`: A 4-D `Tensor` of type `float` and shape `[batch, height, width, in_channels]`.
* `filter`: A 4-D `Tensor` with the same type as `value` and shape
  `[height, width, output_channels, in_channels]`. `filter`'s `in_channels` dimension must
  match that of `value`.
* `output_shape`: A 1-D `Tensor` representing the output shape of the deconvolution op.
* `strides`: A list of `int`. The stride of the sliding window for each dimension of the input tensor.
* `padding`: A string, either `'VALID'` or `'SAME'`. The padding algorithm.
* `name`: Optional name for the returned tensor.

I think it makes sense to follow the same interface design and to adapt some of the names to be consistent with **Theano's** `conv2d` interface. In particular, having users explicitly specify the output shape disambiguates the case where strides are greater than one.

I propose the following interface:

#### `theano.tensor.nnet.conv2d_transpose(input, filters, output_shape, input_shape=None, filter_shape=None, border_mode='valid', subsample=(1, 1), filter_flip=True, filter_dilation=(1, 1))`

* `input`: Symbolic 4D tensor, mini-batch of feature map stacks, of shape
  `(batch size, input channels, input rows, input columns)`.
* `filters`: Symbolic 4D tensor, set of filters used in CNN layer of shape
  `(input channels, output channels, filter rows, filter columns)`. **Note: the order for
  `output channels` and `input channels` is reversed with respect to `conv2d`.**
* `output_shape`: `tuple` representing the shape of the output of `conv2d_transpose`.
* `input_shape`: ...
* `filter_shape`: ...
* `border_mode`: ...
* `subsample`: ...
* `filter_flip`: ...
* `filter_dilation`: ...

**Comment by lamblin** [comment]
It makes sense to me, this function should (as much as possible) be centered on the transposed convolution itself (in terms of what the input and output are, for instance).

I think the following parameters also need some clarifications:
- `subsample`: it means that the original (non-transposed) convolution was doing some subsampling. So `conv2d_transpose` actually does something like "input dilation". TensorFlow names it "strides". I would either keep `subsample` with a warning, or rename it `input_dilation` and say that it corresponds to `subsample` in the non-transposed convolution.
- `border_mode`: what was "padding" for the forward convolution means "cropping" the output of the transposed one. "valid" corresponds to no cropping, "full" to maximal cropping.
